File: src/lib/social/crypto.ts
import crypto from "node:crypto";

const ALGO = "aes-256-gcm";
const IV_LEN = 12;
const TAG_LEN = 16;
const SALT = Buffer.from("pennylime-social-v1", "utf8");

function deriveKey(): Buffer {
  const secret = process.env.NEXTAUTH_SECRET;
  if (!secret) throw new Error("NEXTAUTH_SECRET required for social crypto");
  return crypto.scryptSync(secret, SALT, 32);
}

export function encrypt(plaintext: string): string {
  const key = deriveKey();
  const iv = crypto.randomBytes(IV_LEN);
  const cipher = crypto.createCipheriv(ALGO, key, iv);
  const encrypted = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  // format: base64(iv || tag || ciphertext)
  return Buffer.concat([iv, tag, encrypted]).toString("base64");
}

export function decrypt(payload: string): string {
  const buf = Buffer.from(payload, "base64");
  const iv = buf.subarray(0, IV_LEN);
  const tag = buf.subarray(IV_LEN, IV_LEN + TAG_LEN);
  const ciphertext = buf.subarray(IV_LEN + TAG_LEN);
  const key = deriveKey();
  const decipher = crypto.createDecipheriv(ALGO, key, iv);
  decipher.setAuthTag(tag);
  const decrypted = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
  return decrypted.toString("utf8");
}

File: bot/crypto.py
"""
Python mirror of src/lib/social/crypto.ts.
Same SALT + same scrypt params + same byte layout, so blobs encrypted by
Node.encryptToken can be decrypted here, and vice versa.
"""
import os
import base64
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.scrypt import Scrypt
from cryptography.hazmat.backends import default_backend

SALT = b"pennylime-social-v1"
IV_LEN = 12
TAG_LEN = 16

_cached_key: bytes | None = None


def _key() -> bytes:
    """Derive AES-256 key from NEXTAUTH_SECRET via scrypt. Cached for process lifetime."""
    global _cached_key
    if _cached_key is not None:
        return _cached_key
    secret = os.environ["NEXTAUTH_SECRET"].encode("utf-8")
    # n=16384, r=8, p=1 are Node's scrypt defaults — must match crypto.ts exactly
    kdf = Scrypt(salt=SALT, length=32, n=16384, r=8, p=1, backend=default_backend())
    _cached_key = kdf.derive(secret)
    return _cached_key


def decrypt_token(payload: str) -> str:
    """Decrypt a base64 blob produced by Node's encryptToken."""
    raw = base64.b64decode(payload)
    if len(raw) < IV_LEN + TAG_LEN + 1:
        raise ValueError("crypto: payload too short")
    iv = raw[:IV_LEN]
    tag = raw[IV_LEN : IV_LEN + TAG_LEN]
    ct = raw[IV_LEN + TAG_LEN :]
    aesgcm = AESGCM(_key())
    # AESGCM.decrypt expects ct concatenated with the tag at the end
    plaintext = aesgcm.decrypt(iv, ct + tag, None)
    return plaintext.decode("utf-8")
